rust/src/delta.rs

```diff
@@ -613,6 +662,143 @@ impl std::fmt::Debug for DeltaTable {
+            StorageError::AlreadyExists => DeltaTransactionError::VersionExists,
+            _ => DeltaTransactionError::Storage { source: error },
+        }
+    }
+}
+
+pub struct DeltaTransactionOptions {
+    max_retry_commit_attempts: u32,
+}
+
+const DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS: u32 = 10000000;
+
+pub struct DeltaTransaction<'a> {
+    delta_table: &'a mut DeltaTable,
+    options: DeltaTransactionOptions,
+}
+
+impl<'a> DeltaTransaction<'a> {
+    pub fn new(delta_table: &'a mut DeltaTable, options: Option<DeltaTransactionOptions>) -> Self {
+        DeltaTransaction { 
+            delta_table,
+            options: options.map_or_else(|| DeltaTransactionOptions { 
+                max_retry_commit_attempts: DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS 
+            }, |o| o),
+        }
+    }
+
+    pub async fn commit_all(&mut self, actions: &[Action], _operation: Option<DeltaOperation>) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        // TODO: stubbing `operation` parameter (which will be necessary for writing the CommitInfo action), but leaving it unused for now.
+        // `CommitInfo` is a fairly dynamic data structure so we should work out the data structure approach separately.
+
+        // TODO: calculate isolation level to use when checking for conflicts.
+        // Leaving conflict checking unimplemented for now to get the "single writer" implementation off the ground.
+        // Leaving some commmented code in place as a guidepost for the future.
+
+        // let no_data_changed = actions.iter().all(|a| match a {
+        //     Action::add(x) => !x.dataChange,
+        //     Action::remove(x) => !x.dataChange,
+        //     _ => false,
+        // });
+        // let isolation_level = if no_data_changed {
+        //     IsolationLevel::SnapshotIsolation
+        // } else { 
+        //     IsolationLevel::Serializable
+        // };
+
+        // TODO: create a CommitInfo action and prepend it to actions.
+
+        // Serialize all actions that are part of this log entry.
+        let mut jsons = Vec::<String>::new();
+
+        for action in actions {
+            let json = serde_json::to_string(action)?;
+            jsons.push(json);
+        }
+        
+        let log_entry = jsons.join("\n");
+        let log_entry = log_entry.as_bytes();
+
+        // try to commit in a loop in case other writers write the next version first
+        let version = self.try_commit_loop(log_entry).await?;
+
+        // NOTE: since we have the log entry in memory already, 
+        // we could optimize this further by merging the log entry instead of updating from storage.
+        self.delta_table.update().await?;
+
+        Ok(version)
+    }
+
+    async fn try_commit_loop(&mut self, log_entry: &[u8]) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        let mut attempt_number: u32 = 0;
+
+        loop {
+            let commit_result = self.try_commit(log_entry).await;
```

Comment:
This is offering no means for the caller to figure out why the commit is failing which I imagine is going to lead to some frustration down the road.

@houqp how would you feel about at a minimum logging an error or debug level message about the error attached to `commit_result`?

Reply:
yeah, I think we should break the loop and propagate the error back to caller immediately if the error cannot be resolved by optimistic retry. For example, `storage.put_obj` call in `try_commit` can fail for network or authentication error, which will always fail on retry. Another example would be unresolvable transaction conflicts, which will require the caller discard the transaction entirely and redo the compute from scratch.

For errors like `"the next version you picked is already committed another writer"`, we can safely retry without logging any error because this is expected and normal behavior.

I do think it's valuable to add a debug logging in all cases, whether the error can be resolved by optimistic retry or not.

Lastly, as a future improvement, I think it would be cool to return retry count as part of the result to the caller so that can be collected and sent to whatever monitoring system being used. This way, application owner can keep an eye on whether the system is spending too much time on doing busy retries.

Reply:
@houqp - I have added [this commit](https://github.com/delta-io/delta-rs/pull/84/commits/f46833da900c959dfa93b938d425e79b06842c6b) which fails fast on anything other than an optimistic concurrency retry based on version number and also adds `debug` logging to all commit failure cases. The pattern matching approach I have taken does undo your previous [mention of holding the attempt number check 'til the end](https://github.com/delta-io/delta-rs/pull/84#discussion_r570000729) since pattern match needs to start with most restrictive check first. Let me know you think the single pattern match is worth the trade-off or would prefer additional lines to optimize. I personally tend to prefer fewer code lines when possible.

Reply:
Also - I added a note to potentially include `num_retries` in the error object later per your suggestion.

Reply:
I am also for achieving the same goal with least amount of code :)

This is just my personal preference: I find the logic a little bit easier for to follow with a single match branch and move the max retry check into the branch as a nested if block. Another benefit with single branch is we would never have to worry about restrictiveness of match branches going forward, although I think the compiler will likely complain if an incorrect order results in unreachable code.